In [53]:
import pandas as pd

In [ ]:
df = pd.read_csv('../data/companies.csv', low_memory=False)

df.head()

In [5]:
from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('/content/drive/MyDrive/BTT - G&M1A/G&M/Milestone 1/companies.csv', low_memory=False)
df.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


id  Unnamed: 0.1 entity_type  entity_id  parent_id  \
0      c:1             0     Company          1        NaN   
1     c:10             1     Company         10        NaN   
2    c:100             2     Company        100        NaN   
3  c:10000             3     Company      10000        NaN   
4  c:10001             4     Company      10001        NaN   

                 name     normalized_name                    permalink  \
0            Wetpaint            wetpaint            /company/wetpaint   
1             Flektor             flektor             /company/flektor   
2               There               there               /company/there   
3             MYWEBBO             mywebbo             /company/mywebbo   
4  THE Movie Streamer  the movie streamer  /company/the-movie-streamer   

     category_code     status  ... first_milestone_at last_milestone_at  \
0              web  operating  ...         2010-09-05        2013-09-18   
1      games_video   acquired  ...                NaN               NaN   
2      games_video   acquired  ...         2003-02-01        2011-09-23   
3  network_hosting  operating  ...                NaN               NaN   
4      games_video  operating  ...                NaN               NaN   

  milestones relationships        created_by           created_at  \
0        5.0          17.0  initial-importer  2007-05-25 06:51:27   
1        NaN           6.0  initial-importer  2007-05-31 21:11:51   
2        4.0          12.0  initial-importer  2007-08-06 23:52:45   
3        NaN           NaN               NaN  2008-08-24 16:51:57   
4        NaN           NaN               NaN  2008-08-24 17:10:34   

            updated_at        lat         lng   ROI  
0  2013-04-13 03:29:00  47.606209 -122.332071  15.5  
1  2008-05-23 23:23:14  34.021122 -118.396467   NaN  
2  2013-11-04 02:09:48  37.562992 -122.325525   NaN  
3  2008-09-06 14:19:18        NaN         NaN   NaN  
4  2008-09-06 14:19:18        NaN         NaN   NaN  

[5 rows x 44 columns]

In [6]:
df.shape

(196553, 44)

In [7]:
df.shape

(196553, 44)

In [ ]:
df.dtypes

id                      object
Unnamed: 0.1             int64
entity_type             object
entity_id                int64
parent_id              float64
name                    object
normalized_name         object
permalink               object
category_code           object
status                  object
founded_at              object
closed_at               object
domain                  object
homepage_url            object
twitter_username        object
logo_url                object
logo_width             float64
logo_height            float64
short_description       object
description             object
overview                object
tag_list                object
country_code            object
state_code              object
city                    object
region                  object
first_investment_at     object
last_investment_at      object
investment_rounds      float64
invested_companies     float64
first_funding_at        object
last_funding_at         object
funding_

In [ ]:
df.isnull().sum()

id                          0
Unnamed: 0.1                0
entity_type                 0
entity_id                   0
parent_id              196553
name                       23
normalized_name            26
permalink                   0
category_code           73367
status                      0
founded_at             105326
closed_at              193933
domain                  70008
homepage_url            70008
twitter_username       115962
logo_url                86443
logo_width              86443
logo_height             86443
short_description      189422
description            104505
overview                69582
tag_list               115101
country_code           108563
state_code             145650
city                   112663
region                      0
first_investment_at    193970
last_investment_at     193970
investment_rounds      193962
invested_companies     193962
first_funding_at       165046
last_funding_at        165046
funding_rounds         164846
funding_to

In [ ]:
df['status'].unique()

array(['operating', 'acquired', 'closed', 'ipo'], dtype=object)

In [ ]:
df['category_code'].unique()

array(['web', 'games_video', 'network_hosting', 'advertising',
       'cleantech', nan, 'enterprise', 'other', 'consulting', 'mobile',
       'health', 'software', 'analytics', 'finance', 'education',
       'medical', 'manufacturing', 'biotech', 'ecommerce',
       'public_relations', 'hardware', 'search', 'news', 'government',
       'security', 'photo_video', 'travel', 'semiconductor', 'social',
       'legal', 'transportation', 'hospitality', 'sports', 'nonprofit',
       'fashion', 'messaging', 'music', 'automotive', 'design',
       'real_estate', 'local', 'nanotech', 'pets'], dtype=object)

In [ ]:
df['country_code'].unique()

array(['USA', nan, 'MAR', 'IND', 'AUS', 'FRA', 'JPN', 'NLD', 'EGY', 'ISR',
       'GBR', 'THA', 'CAN', 'AUT', 'IRL', 'SWE', 'DEU', 'BRA', 'FIN',
       'RUS', 'SGP', 'MEX', 'CHN', 'ESP', 'ISL', 'KOR', 'TUR', 'DNK',
       'ARG', 'PAK', 'HUN', 'POL', 'GRC', 'PRT', 'BLR', 'CSS', 'MKD',
       'CHE', 'SVN', 'UKR', 'ITA', 'NZL', 'LIE', 'NOR', 'CZE', 'VNM',
       'HRV', 'BEN', 'CHL', 'GHA', 'ZAF', 'MYS', 'EST', 'BEL', 'SVK',
       'TWN', 'CRI', 'HKG', 'BGD', 'BOL', 'LBN', 'LUX', 'AZE', 'COL',
       'PHL', 'ARE', 'IDN', 'ROM', 'ANT', 'CYM', 'NGA', 'GIN', 'CYP',
       'LKA', 'SWZ', 'VGB', 'MLT', 'SAU', 'KEN', 'BGR', 'PER', 'LVA',
       'LAO', 'NPL', 'MDA', 'CMR', 'UGA', 'TUN', 'URY', 'MUS', 'VEN',
       'OMN', 'ECU', 'KWT', 'JOR', 'LTU', 'ALB', 'QAT', 'UMI', 'PST',
       'REU', 'GTM', 'PCN', 'BHS', 'GEO', 'BWA', 'DZA', 'GRD', 'ZMB',
       'GIB', 'PAN', 'MTQ', 'PRI', 'BIH', 'DMA', 'BHR', 'SYC', 'SLE',
       'TTO', 'VCT', 'ARA', 'BRB', 'NAM', 'SLV', 'AFG', 'BLZ', 'UZB',
       'IRN', '

In [ ]:
df['state_code'].unique()

array(['WA', 'CA', nan, 'NY', 'NM', 'TX', 'OH', 'NJ', 'IL', 'MA', 'NC',
       'CT', 'PA', 'NH', 'MI', 'AZ', 'TN', 'GA', 'FL', 'MD', 'UT', 'OR',
       'MO', 'AR', 'CO', 'KS', 'MN', 'NV', 'ID', 'IN', 'IA', 'VA', 'KY',
       'LA', 'HI', 'WV', 'WI', 'SC', 'NE', 'DE', 'DC', 'AL', 'VT', 'RI',
       'OK', 'ME', 'MS', 'MT', 'SD', 'ND', 'AK', 'WY'], dtype=object)

In [ ]:
df['region'].unique()

array(['Seattle', 'Los Angeles', 'SF Bay', ..., 'Parnell Auckland',
       'Jalandhar City', 'D.I.Khan'], shape=(5847,), dtype=object)

In [ ]:
#Status for rows where investment columns are empty
df.loc[df['investment_rounds'].isna(), 'status'].value_counts()


status
operating    181188
acquired       9281
closed         2573
ipo             920
Name: count, dtype: int64

In [ ]:
#Status for rows where funding amount columns are empty
df.loc[df['funding_total_usd'].isna(), 'status'].value_counts()

status
operating    160130
acquired       7059
closed          836
ipo             654
Name: count, dtype: int64

# 1. Under-Capitalized Population Identification (2% Threshold)

## Step 1: Calculate Funding Distribution Percentiles

```
ALGORITHM: Find 2% Funding Threshold
1. Calculate percentiles [1%, 2%, 5%, 10%] of funding_total_usd
2. Display percentile values for reference  
3. Extract 2% percentile as threshold value
4. Format and display threshold amount

EXPECTED OUTPUT:
- Percentile table showing funding distribution
- 2% threshold dollar amount (e.g., "$50,000")
```

Have a looksee at pd.DataFrame.quantile() for percentile calculations.

## Step 2: Identify Under-Capitalized Companies

```
ALGORITHM: Create Under-Capitalized Flag
1. Create boolean flag where:
   - funding_total_usd <= 2% threshold OR
   - funding_total_usd is null OR  
   - funding_total_usd equals zero
2. Count total under-capitalized companies
3. Calculate percentage of total dataset
4. Display summary statistics

EXPECTED OUTPUT:
- under_capitalized column added to dataframe
- Count: "X,XXX companies (XX.X%)"
```

Check out pandas conditional selection for flag creation.

## Step 3: Compare Populations

```
ALGORITHM: Population Comparison Analysis
1. Split data into under_capitalized vs well_funded groups
2. Generate descriptive statistics for each group
3. Compare key metrics (mean, median, std, min, max)
4. Identify funding gaps between populations

EXPECTED OUTPUT:
- Side-by-side statistical comparison
- Key differences highlighted
```

Use pandas groupby() and describe() methods for comparative statistics.

## Code for Step 1

In [19]:
# Percentiles
percentiles = df['funding_total_usd'].quantile([0.01, 0.02, 0.05, 0.10])

percentiles = percentiles.reset_index()
percentiles.columns = ['Percentile', 'Funding (USD)']

percentiles['Funding (USD)'] = percentiles['Funding (USD)'].apply(lambda x: f"${x:,.0f}")
percentiles['Percentile'] = (percentiles['Percentile']*100).astype(int).astype(str) + '%'

percentiles



Percentile Funding (USD)
0         1%       $11,700
1         2%       $19,616
2         5%       $40,000
3        10%      $100,000

In [18]:
#2% threshold
two_percent = df['funding_total_usd'].quantile(0.02)

print(f"2% funding threshold: ${two_percent:,.0f}")

2% funding threshold: $19,616


## Code for Step 2

In [20]:
#flag for under-capitalized companies
df['under_capitalized'] = (
    (df['funding_total_usd'] <= two_percent) |
    (df['funding_total_usd'].isna()) |
    (df['funding_total_usd'] == 0))

df[['funding_total_usd', 'under_capitalized']].head()


funding_total_usd  under_capitalized
0         39750000.0              False
1                NaN               True
2                NaN               True
3                NaN               True
4                NaN               True

In [27]:
# total under-capitalized companies

under_cap_count = df['under_capitalized'].sum()
under_cap_percent = (under_cap_count / len(df)) * 100

print(f"Under-capitalized companies: {under_cap_count:,} ({under_cap_percent:.1f}%)")


Under-capitalized companies: 169,238 (86.1%)


In [28]:
# Summary statistics for under_capitalized column
df['under_capitalized'].value_counts(dropna=False)

under_capitalized
True     169238
False     27315
Name: count, dtype: int64

## Code for Step 3

In [29]:
# Split data into under-capitalized vs well-funded groups
under_cap = df[df['under_capitalized'] == True]
well_funded = df[df['under_capitalized'] == False]

print(f"Under-capitalized rows: {len(under_cap):,}")
print(f"Well-funded rows: {len(well_funded):,}")

Under-capitalized rows: 169,238
Well-funded rows: 27,315


In [31]:
# Generate descriptive statistics for each group
under_cap_stats = under_cap['funding_total_usd'].describe()
well_funded_stats = well_funded['funding_total_usd'].describe()

print(f"Under-capitalized stats:\n{under_cap_stats}\n")
print(f"Well-funded stats:\n{well_funded_stats}")

Under-capitalized stats:
count      559.000000
mean     11530.096601
std       5392.753633
min        291.000000
25%       8000.000000
50%      11700.000000
75%      15000.000000
max      19616.000000
Name: funding_total_usd, dtype: float64

Well-funded stats:
count    2.731500e+04
mean     1.511950e+07
std      6.841578e+07
min      1.965400e+04
25%      6.000000e+05
50%      2.775000e+06
75%      1.130000e+07
max      5.700000e+09
Name: funding_total_usd, dtype: float64


In [32]:
# Compare key metrics and identify funding gaps between groups

group_stats = df.groupby('under_capitalized')['funding_total_usd'].describe()[['mean', '50%', 'std', 'min', 'max']]
group_stats.rename(index={True: 'Under-Capitalized', False: 'Well-Funded'}, inplace=True)
group_stats.rename(columns={'50%': 'median'}, inplace=True)

group_stats

mean     median           std      min  \
under_capitalized                                                   
Well-Funded        1.511950e+07  2775000.0  6.841578e+07  19654.0   
Under-Capitalized  1.153010e+04    11700.0  5.392754e+03    291.0   

                            max  
under_capitalized                
Well-Funded        5.700000e+09  
Under-Capitalized  1.961600e+04

# 2. Success/Failure Proxy Analysis

## Step 1: Analyze Status Distribution

```
ALGORITHM: Status Category Analysis
1. Count frequency of each status value
2. Calculate percentage distribution
3. Display formatted status breakdown
4. Identify most/least common statuses

EXPECTED OUTPUT:
- Status distribution table with counts and percentages
- Key patterns identified (e.g., "operating" dominance)
```

value_counts() for the win here.

## Step 2: Create Success/Failure Categories

```
ALGORITHM: Success Proxy Variable Creation
1. Define category mappings:
   - SUCCESS: ['acquired', 'ipo']
   - FAILURE: ['closed']
   - ACTIVE: ['operating']
   - UNKNOWN: all others
2. Apply mapping to create success_category column
3. Validate mapping completeness
4. Display new category distribution

EXPECTED OUTPUT:
- success_category column added
- Clear success/failure/active/unknown breakdown
```

Take advantage of pandas .map() or .apply() functions

### Step 3: Calculate Success Rates by Population

```
ALGORITHM: Success Rate Analysis
1. Calculate overall success rate (success / (success + failure))
2. Filter to under-capitalized companies → calculate success rate
3. Filter to well-funded companies → calculate success rate  
4. Compare rates between populations
5. Statistical significance assessment

EXPECTED OUTPUT:
- Overall Success Rate: XX.X%
- Under-cap Success Rate: XX.X%
- Well-funded Success Rate: XX.X%
- Success gap analysis
```

Use pandas groupby() and aggregation functions to compute success rates

## Code for Step 1

In [39]:
# Status distribution (counts + percentages)
status_counts = df['status'].value_counts(dropna=False)
status_percent = (df['status'].value_counts(normalize=True, dropna=False) * 100).round(2)

status_dist = pd.DataFrame({'count': status_counts, 'percent': status_percent})
status_dist = status_dist.sort_values('count', ascending=False)

status_dist


count  percent
status                    
operating  183441    93.33
acquired     9394     4.78
closed       2584     1.31
ipo          1134     0.58

In [38]:
# Identify most and least common statuses
most_common = status_dist.index[0]
least_common = status_dist.index[3]

print(f"Most common status: {most_common}")
print(f"Least common status: {least_common}")

Most common status: operating
Least common status: ipo


## Code for Step 2

In [42]:
# category mappings
status_map = {
    'acquired': 'SUCCESS',
    'ipo': 'SUCCESS',
    'closed': 'FAILURE',
    'operating': 'ACTIVE'}

# Others = UNKNOWN
df['success_category'] = df['status'].map(status_map).fillna('UNKNOWN')

df[['status', 'success_category']].head()


status success_category
0  operating           ACTIVE
1   acquired          SUCCESS
2   acquired          SUCCESS
3  operating           ACTIVE
4  operating           ACTIVE

In [44]:
#Validate mapping completeness
df['success_category'].isna().sum()

np.int64(0)

In [46]:
# Display new category distribution
df['success_category'].value_counts()

success_category
ACTIVE     183441
SUCCESS     10528
FAILURE      2584
Name: count, dtype: int64

## Code for Step 2

In [48]:
# Calculate overall success rate
total_success = (df['success_category'] == 'SUCCESS').sum()
total_failure = (df['success_category'] == 'FAILURE').sum()

overall_success_rate = total_success / (total_success + total_failure)
print(f"Overall Success Rate: {overall_success_rate:.2%}")

Overall Success Rate: 80.29%


In [49]:
# Success rate for under-capitalized companies
under_capitalized  = df[df['under_capitalized'] == True]

under_success = (under_capitalized ['success_category'] == 'SUCCESS').sum()
under_failure = (under_capitalized ['success_category'] == 'FAILURE').sum()

under_success_rate = under_success / (under_success + under_failure)
print(f"Under-Capitalized Success Rate: {under_success_rate:.2%}")

Under-Capitalized Success Rate: 89.22%


In [50]:
# Success rate for well-funded companies
well_funded = df[df['under_capitalized'] == False]

well_success = (well_funded['success_category'] == 'SUCCESS').sum()
well_failure = (well_funded['success_category'] == 'FAILURE').sum()

well_success_rate = well_success / (well_success + well_failure)
print(f"Well-Funded Success Rate: {well_success_rate:.2%}")

Well-Funded Success Rate: 62.90%


In [54]:
# Compare rates between populations
gap = under_success_rate - well_success_rate

print(f"Overall Success Rate: {overall_success_rate:.2%}")
print(f"Under-Capitalized Success Rate: {under_success_rate:.2%}")
print(f"Well-Funded Success Rate: {well_success_rate:.2%}")
print(f"Success Rate Gap: {gap:.2%}")

Overall Success Rate: 80.29%
Under-Capitalized Success Rate: 89.22%
Well-Funded Success Rate: 62.90%
Success Rate Gap: 26.32%


The overall success rate was 80.29%.
Under-capitalized companies had a success rate of 89.22%, and well-funded companies had 62.90%.
That’s a 26.32% gap between the two groups.
This means in our data, under-capitalized startups were more likely to succeed.


The higher success rate for under-capitalized companies might be because many of them have missing or zero funding values, and some of those companies still show “acquired” or “IPO” statuses.

# 3. Geographic Patterns for Heatmap Prep

## Step 1: Analyze Geographic Distribution

```
ALGORITHM: Geographic Coverage Analysis
1. Count companies by country_code (focus on 'usa')
2. Calculate US company percentage of total dataset
3. Identify international vs domestic distribution
4. Assess geographic data completeness

EXPECTED OUTPUT:
- Top 10 countries by startup count
- US market dominance percentage
- International market insights
```

Use pandas groupby() and aggregation functions to compute geographic distributions

## Step 2: US State Analysis

```
ALGORITHM: State-Level Distribution Analysis
1. Filter dataset to US companies only
2. Count companies by state_code
3. Rank states by startup concentration
4. Calculate missing state data percentage
5. Identify data quality issues

EXPECTED OUTPUT:
- Top 15 states ranked by company count
- State coverage completeness metrics
- Missing data patterns identified
```

## Step 3: Geographic Success Patterns

```
ALGORITHM: State Success Rate Analysis
1. FOR each top 10 state:
   a. Filter companies in that state
   b. Calculate success rate (success/(success+failure))
   c. Calculate under-capitalized percentage  
   d. Assess sample size adequacy
2. Rank states by success metrics
3. Identify geographic opportunity patterns

EXPECTED OUTPUT:
- State-by-state success rate table
- Under-cap concentration by geography
- Geographic opportunity hotspots identified
```

## Step 4: Geographic Data Quality Assessment

```
ALGORITHM: Data Standardization Assessment
1. Examine state_code format consistency
2. Identify non-standard state codes (length ≠ 2)
3. Check for full state names vs abbreviations
4. Assess region data completeness
5. Document cleaning requirements

EXPECTED OUTPUT:
- Data quality issues inventory
- Standardization requirements list
- Geographic coverage gaps identified
```

# 4. Industry Growth Preliminary Analysis

## Step 1: Industry Distribution Analysis

```
ALGORITHM: Industry Concentration Assessment
1. Count companies by category_code
2. Rank top 20 industries by company count
3. Calculate market concentration metrics:
   - Top 5 industries market share
   - Top 10 industries market share  
4. Assess industry diversity vs concentration

EXPECTED OUTPUT:
- Industry ranking table (top 20)
- Market concentration percentages
- Industry diversity assessment
```

## Step 2: Industry Success Patterns

```
ALGORITHM: Industry Performance Analysis
1. FOR each top 15 industry:
   a. Count total companies
   b. Calculate success rate (where data available)
   c. Calculate under-capitalized percentage
   d. Calculate median funding amount
   e. Store metrics in analysis array
2. Sort industries by success rate
3. Identify high-performing vs struggling sectors

EXPECTED OUTPUT:  
- Industry performance scorecard
- Success rate rankings by sector
- Under-cap opportunity identification
```

## Step 3: Under-Capitalized Industry Opportunities

```
ALGORITHM: Opportunity Sector Identification
1. Count under-cap companies by industry
2. Rank industries by under-cap population size
3. Filter for "mixed success" industries (20-70% success rate, 50+ companies)
4. Cross-reference high under-cap populations with mixed success
5. Identify prime opportunity sectors for investment

EXPECTED OUTPUT:
- Industries with most under-cap companies (top 15)
- Mixed-success industries list (risk profiling candidates)
- Investment opportunity matrix
```

## Step 4: Industry Data Quality Assessment

```
ALGORITHM: Category Consolidation Assessment
1. Count unique category_code values
2. Identify categories with <10 companies (consolidation candidates)
3. Identify categories with <5 companies (merge required)  
4. Sample small categories for consolidation planning
5. Assess semantic similarity for grouping

EXPECTED OUTPUT:
- Total category count and fragmentation metrics
- Small category consolidation list
- Category grouping recommendations
```

# 5. Data Quality

## Step 1: Overall Data Quality Assessment

```
ALGORITHM: Missing Value Impact Analysis
1. FOR each key column [funding_total_usd, status, country_code, state_code, category_code]:
   a. Count missing values
   b. Calculate missing percentage
   c. Assess impact on analysis capabilities
2. Calculate US company data completeness rate
3. Identify critical data gaps for project goals

EXPECTED OUTPUT:
- Missing value report by column
- US data completeness percentage  
- Critical data gap identification
```

I know you have a doc inside of your google docs that has the data quality findings documented. Please transfer those findings here. This can be in a markdown file or in a notebook cell, but they should be here!

## Step 2: Key Insights Summary

```
ALGORITHM: Findings Documentation
1. Summarize under-capitalized population insights:
   - Total count and percentage
   - Success rate differential
2. Document geographic opportunity findings:
   - Top startup concentration states
   - Success rate patterns by geography  
3. Catalog industry opportunity insights:
   - Mixed-success industries for risk profiling
   - High under-cap concentration sectors
4. Compile data quality findings

EXPECTED OUTPUT:
- Executive summary of key findings
- Quantified insights for stakeholder communication
- Opportunity identification for next phases
```